src/test/test-Stream.ts

```diff
@@ -2084,4 +2085,312 @@ describe("Stream", () => {
+					value: 2,
+				},
+				{
+					value: 3,
+				},
+			];
+
+			source.writeEach(() => sourceData.shift());
+
+			let isAborted = false;
+
+			source.aborted().catch(() => {
+				isAborted = true;
+			});
+
+			const batched = pipeWithDelay(source).batch(2);
+
+			const delayedProcessing = resolveBatchToAsyncValues(batched);
+
+			const dest = await delayedProcessing;
+			expect(dest).to.deep.equal([[1, 2]]);
+
+			expect(isAborted).to.equal(true);
+		});
+
+		it("waits for source stream to end", async () => {
+			const d = defer();
+			const slowEndingSource = s.transform<number>(
+				(readable, writable) => {
+					readable.forEach(
+						(v) => writable.write(v),
+						(error?: Error) => {
+							writable.end(error, readable.result());
+							return d.promise;
+						}
+					);
+				}
+			);
+			const writes = [
+				track(s.write(1)),
+				track(s.write(2)),
+				track(s.end()),
+			];
+
+			const batched = slowEndingSource.batch(1);
+			const mres = track(batched.result());
+			readInto(batched, batchResults);
+
+			await settle([writes[0].promise, writes[1].promise]);
+			await delay(1);
+			expect(batchResults).to.deep.equal([[1], [2]]);
+			expect(writes[0].isFulfilled).to.equal(true);
+			expect(writes[1].isFulfilled).to.equal(true);
+			expect(writes[2].isFulfilled).to.equal(false);
+			expect(mres.isFulfilled).to.equal(false);
+
+			d.resolve();
+			await settle([mres.promise]);
+			expect(writes[2].isFulfilled).to.equal(true);
+		});
+
+		it("waits for destination stream to end", async () => {
+			const d = defer();
+			const slowEnder: Transform<number[], number[]> = (
+				readable,
+				writable
+			) => {
+				readable.forEach(
+					(v) => writable.write(v),
+					(error?: Error) => {
+						writable.end(error, readable.result());
+						return d.promise;
+					}
+				);
+			};
+			const w1 = track(s.write(1));
+			const w2 = track(s.write(2));
+			const we = track(s.end());
+
+			const batched = s.batch(1);
+			const mres = track(batched.result());
+			const slowed = batched.transform(slowEnder);
+			const sres = track(slowed.result());
+			readInto(slowed, batchResults);
+
+			await delay(1);
+			expect(batchResults).to.deep.equal([[1], [2]]);
+			expect(w1.isFulfilled).to.equal(true);
+			expect(w2.isFulfilled).to.equal(true);
+			expect(we.isFulfilled).to.equal(false);
+			expect(mres.isFulfilled).to.equal(false);
+			expect(sres.isFulfilled).to.equal(false);
+
+			d.resolve();
+			await settle([mres.promise, sres.promise]);
+		});
+
+		it("aborts from source to sink", async () => {
+			const sink = s.batch(1).map(identity);
+			const ab = track(sink.aborted());
+			s.abort(abortError);
+			await settle([ab.promise]);
+			expect(ab.reason).to.equal(abortError);
+		});
+
+		it("aborts from sink to source", async () => {
+			const ab = track(s.aborted());
+			const sink = s.batch(1).map(identity);
+			sink.abort(abortError);
+			await settle([ab.promise]);
+			expect(ab.reason).to.equal(abortError);
+		});
```

Comment:
Please also add tests the check when downstream throws errors, in all kinds of batches (minSize, maxSize and on timeout).

Reply:
Tests added